src/virtio-blk.c

```diff
@@ -0,0 +1,196 @@
+}
+
+static ssize_t virtio_blk_write(struct virtio_blk_dev *dev,
+                                void *data,
+                                off_t offset,
+                                size_t size)
+{
+    return diskimg_write(dev->diskimg, data, offset, size);
+}
+
+static ssize_t virtio_blk_read(struct virtio_blk_dev *dev,
+                               void *data,
+                               off_t offset,
+                               size_t size)
+{
+    return diskimg_read(dev->diskimg, data, offset, size);
+}
+
+static void virtio_blk_complete_request(struct virtq *vq)
+{
+    struct virtio_blk_dev *dev = (struct virtio_blk_dev *) vq->dev;
+    vm_t *v = container_of(dev, vm_t, virtio_blk_dev);
+    uint8_t status;
+    struct vring_packed_desc *desc;
+    struct virtio_blk_req req;
+
+    while ((desc = virtq_get_avail(vq))) {
+        struct vring_packed_desc *used_desc = desc;
+        int r = 0;
+
+        memcpy(&req, vm_guest_to_host(v, (void *) desc->addr), desc->len);
+        if (req.type == VIRTIO_BLK_T_IN || req.type == VIRTIO_BLK_T_OUT) {
+            if (!virtq_check_next(desc))
+                return;
+            desc = virtq_get_avail(vq);
+            req.data_size = desc->len;
+            req.data = vm_guest_to_host(v, (void *) desc->addr);
+
+            ssize_t r;
+            if (req.type == VIRTIO_BLK_T_IN)
+                r = virtio_blk_read(dev, req.data, req.sector << 9,
+                                    req.data_size);
+            else
+                r = virtio_blk_write(dev, req.data, req.sector << 9,
+                                     req.data_size);
+
+            status = r < 0 ? VIRTIO_BLK_S_IOERR : VIRTIO_BLK_S_OK;
+        } else {
+            status = VIRTIO_BLK_S_UNSUPP;
+        }
+        if (!virtq_check_next(desc))
+            return;
+        desc = virtq_get_avail(vq);
+        req.status = vm_guest_to_host(v, (void *) desc->addr);
+        *req.status = status;
+        used_desc->flags ^= (1ULL << VRING_PACKED_DESC_F_USED);
+        used_desc->len = r;
+        dev->virtio_pci_dev.config.isr_cap.isr_status |= VIRTIO_PCI_ISR_QUEUE;
+    }
+}
+
+static struct virtq_ops ops = {
+    .enable_vq = virtio_blk_enable_vq,
+    .complete_request = virtio_blk_complete_request,
+    .notify_used = virtio_blk_notify_used,
+};
+
+void virtio_blk_enable(struct virtio_blk_dev *dev, struct diskimg *diskimg)
+{
+    dev->diskimg = diskimg;
+    dev->config.capacity = diskimg->size >> 9;
+    dev->enable = true;
+    virtio_pci_enable(&dev->virtio_pci_dev);
+    pthread_create(&dev->worker_thread, NULL, (void *) virtio_blk_thread,
+                   (void *) dev);
+}
+
+void virtio_blk_init_pci(struct virtio_blk_dev *virtio_blk_dev,
+                         struct pci *pci,
+                         struct bus *io_bus,
+                         struct bus *mmio_bus)
+{
+    virtio_blk_init(virtio_blk_dev);
+    struct virtio_pci_dev *dev = &virtio_blk_dev->virtio_pci_dev;
```

Comment:
How about making ` virtio_blk_init` and `virtio_blk_enable` become static,  and do `virtio_blk_enable` in `virtio_blk_init_pci` instead of seperate them in `vm_load_diskimg`?

So it will be clear that  `virtio_blk_exit` should only be run when we did  `virtio_blk_init_pci` before.

Reply:
But the value of `enable` in `virtio_blk` isn't specified if we haven't called `virtio_blk_init_pci`